design-proposals/klip-19-materialize-views.md

```diff
@@ -0,0 +1,223 @@
+* `INSERT VALUES` semantics fixed to only allow inserts into mutable `TABLE`s and `STREAM`s, not `VIEW`s.
+* CLI or server enhancements to return useful error message on any statements that are no longer supported. 
+
+TBD:
+* Maintain or drop `CREATE STREAM AS SELECT` functionality?
+ 
+## What is not in scope
+
+* No `VIEW` functionality, only `MATERIALIZED VIEW`s. `VIEW`s have the potential of being useful, as they would allow
+  user to define multiple statements that can later be materialized in a single KS topology. But such work is outside the 
+  scope of this KLIP.
+* New `CREATE TABLE AS SELECT` functionality. This can be done later.
+
+## Value/Return
+
+Differentiating between mutable and read-only data sources will make KSQL easier for both engineers and end-users
+to reason about, the SQL more explicit and is needed to to fix the inconsistencies in `INSERT VALUES` functionality.  
+  
+## Public APIS
+
+We propose the following SQL syntax and associated functionality:
+
+### Tables:
+
+* Add `CREATE MATERIALIZED VIEW` to import an existing topic containing a table. 
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` *not* supported in `WITH` clause.
+* Re-purpose `CREATE TABLE` to create a mutable table backed by a new or existing topic.
+    * `INSERT`s into the table will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of any new topic. 
+* Replace `CREATE TABLE AS SELECT` with `CREATE MATERIALIZED VIEW AS SELECT`.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.   
+* Remove `CREATE TABLE AS SELECT`.
+    
+For those worried about the excessive typing required for `MATERIALIZED VIEW`, we could add a shorthand `CREATE MV` etc. 
+
+### Streams:
+
+Standard SQL already has `MATERIALIZED VIEW`, but has no stream equivalent. 
+We propose a stream equivalent of `MATERIALIZED STREAM VIEW`.
+
+* Add `CREATE MATERIALIZED STREAM VIEW` to import an existing topic containing a stream.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` *not* supported in `WITH` clause.
+* Re-purpose `CREATE STREAM` to create a mutable stream backed by a new or existing topic.
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of any new topic. 
+* Replace `CREATE STREAM AS SELECT` with `CREATE MATERIALIZED STREAM VIEW AS SELECT`.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.   
+* Either `CREATE STREAM AS SELECT` will be removed, or maintained as creating a mutable derived stream. 
```

Comment:
IMHO, keeping `CSAS` is useful.

Reply:
This may be true. Looking for peoples thoughts on this. Thanks for your input.

Reply:
I'd favour removing it unless it's something that's going to be useful to most users. Smaller, simpler API = less user confusion = better user experience = more adoption :)

Reply:
Great, so one vote either way. Perfect! ;) :p

Reply:
If `CSAS` is removed one would have to create a new 'insertable' stream and `UNION` it with the old stream to achieve the same result which would require 2 new streams instead of one, right?
For performance reasons it may be beneficial to keep `CSAS`, but you can judge that way better than I could.

Reply:
I feel that we only need to distinguish "creating a stream / table" from "creating a read-only stream / table" here. But instead of making the newly introduced keyword as "opt-in" (i.e. without the keyword the stream / table would be mutable) maybe it's better making them "opt-out" (i.e. by default they are immutable). The rationale I have is that in practice, most streams / tables would actually be immutable, and perhaps only the source streams need to be mutable. And since not many people have been using `INSERT INTO` in practice so maybe it's okay to break this semantic compatibility:

```
CREATE STREAM AS SELECT   // create an un-insertable stream
CREATE TABLE AS SELECT   // create an un-insertable table
CREATE MUTABLE STREAM AS SELECT     // create un-insertable stream
CREATE MUTABLE TABLE AS SELECT     // create un-insertable table
```

Admittedly, and for tables the `immutable` keyword has a more commonly used alternative `materialized view`, so if we decide to go with `opt-in` I'm also fine.

Reply:
Also a key question we've not yet addressed is, if we ever wants to make ksqlDB to be a real RDBMS, because otherwise we could say we exclude supporting mutable tables at all (i.e. there's only materialized views, in which KLIP-11's proposal seems more reasonable to me).

If we decided to still support direct-inserts-into-table in the long run, I think we should still make this distinguishment --- whatever calling it materialized views or immutable or read-only --- as early as possible.

Reply:
> And since not many people have been using INSERT INTO in practice

How do you/we know this?

About your proposal to flip the default from mutable to immutable: I am not a fan of it, because I think we should try to stick with standard SQL if possible. Additionally, if we remove CT syntax completely, and only have CMV syntax, it's pretty clear to users what going on.

Also note that `INSERT INTO table` is already supported. It's implemented as an append to the corresponding topic (ie, it's an async write; there is no read-after-write consistency). I assume that this feature would be "dropped" if we switch all statements from CT to CMV? \cc @big-andy-coates

Reply:
Yeah that's a fair point. My rationale is just that: in the future do we expect majority of TABLEs to be immutable or mutable, and the language design should be best making the majority case with less keywords if we are designing it for the first time from scratch.

My bet is that immutable tables would be majority, and hence they should enjoy the CT statement with less keywords while the mutable CT should use more keywords. Of course, that's assuming we are re-designing the language from scratch but that's not the case, today effectively mutable CT uses less words, so the question is whether it's worth-while to break this semantic compatibility for the sake of the above argument: if not too many people have already used `INSERT INTO` maybe it's worthwhile, but admittedly I'm do not know statistically how many are there indeed.

Reply:
Well, SQL has already a statement to create immutable tables: `CREATE MATERIALIZE VIEW` -- thus, IMHO we should just use what is already there. Why "flipping" it around introduce non-standard SQL while SQL offers everything we need already?
```
// standard SQL
CREATE TABLE... // mutable
CREATE MATERIALIZED VIEW... // immutable

// why is this better?
CREATE MUTABLE TABLE... // mutable
CREATE TABLE... // immutable
```

If we assume that most people want/need immutable tables, it seems sufficient to educate them to use CMV statements. No need to introduce new syntax and no need to change the semantics of the well established CT statement (I would expect that this would be rather confusing to users).

Reply:
Yeah I think the existing syntax in SQL is a strong one compared to merely saving some keywords, I'm happy to fold my suggestion about flipping it